lightning/src/routing/scoring.rs

```diff
@@ -460,25 +466,421 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	fn set_max_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.max_liquidity_offset_msat = self.capacity_msat.checked_sub(amount_msat).unwrap_or(0);
+		*self.min_liquidity_offset_msat = if amount_msat < self.min_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> Score for ProbabilisticScorerUsingTime<G, T> {
+	#[allow(clippy::float_cmp)]
+	fn channel_penalty_msat(
+		&self, short_channel_id: u64, amount_msat: u64, capacity_msat: u64, source: &NodeId,
+		target: &NodeId
+	) -> u64 {
+		let liquidity_penalty_multiplier_msat = self.params.liquidity_penalty_multiplier_msat;
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let success_probability = self.channel_liquidities
+			.get(&short_channel_id)
+			.unwrap_or(&ChannelLiquidity::new())
+			.as_directed(source, target, capacity_msat, liquidity_offset_half_life)
+			.success_probability(amount_msat);
+		if success_probability == 0.0 {
+			u64::max_value()
+		} else if success_probability == 1.0 {
+			0
+		} else {
+			(-(success_probability.log10()) * liquidity_penalty_multiplier_msat as f64) as u64
+		}
+	}
+
+	fn payment_path_failed(&mut self, path: &[&RouteHop], short_channel_id: u64) {
+		let amount_msat = path.split_last().map(|(hop, _)| hop.fee_msat).unwrap_or(0);
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let network_graph = self.network_graph.read_only();
+		let hop_sources = path.iter().map(|hop| NodeId::from_pubkey(&hop.pubkey));
+		for (source, hop) in hop_sources.zip(path.iter().skip(1)) {
+			let target = NodeId::from_pubkey(&hop.pubkey);
+			let capacity_msat = network_graph.channels()
+				.get(&hop.short_channel_id)
+				.and_then(|channel| channel.as_directed_to(&target).map(|d| d.effective_capacity()))
+				.unwrap_or(EffectiveCapacity::Unknown)
+				.as_msat();
+
+			if hop.short_channel_id == short_channel_id {
+				self.channel_liquidities
+					.entry(hop.short_channel_id)
+					.or_insert_with(ChannelLiquidity::new)
+					.as_directed_mut(&source, &target, capacity_msat, liquidity_offset_half_life)
+					.failed_at_channel(amount_msat);
+				break;
+			}
+
+			self.channel_liquidities
+				.entry(hop.short_channel_id)
+				.or_insert_with(ChannelLiquidity::new)
+				.as_directed_mut(&source, &target, capacity_msat, liquidity_offset_half_life)
+				.failed_downstream(amount_msat);
+		}
+	}
+
+	fn payment_path_successful(&mut self, path: &[&RouteHop]) {
+		let amount_msat = path.split_last().map(|(hop, _)| hop.fee_msat).unwrap_or(0);
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let network_graph = self.network_graph.read_only();
+		let hop_sources = path.iter().map(|hop| NodeId::from_pubkey(&hop.pubkey));
+		for (source, hop) in hop_sources.zip(path.iter().skip(1)) {
```

Comment:
nit: I think a comment for skipping the first node would be good (do we document somewhere that we don't score our own channels?)